File: scripts/scan.js
#!/usr/bin/env node
// Standalone screener scanner - runs outside Next.js
// Usage: node scripts/scan.js [subset_count]
// Uses chart endpoint first (no auth needed), then enriches with fundamentals

const https = require('https');
const fs = require('fs');
const path = require('path');

const UA = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36';
const DATA_FILE = path.join(__dirname, '..', 'data', 'screener_cache.json');

const dataDir = path.join(__dirname, '..', 'data');
if (!fs.existsSync(dataDir)) fs.mkdirSync(dataDir, { recursive: true });

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

function httpsGet(url, headers = {}) {
    return new Promise((resolve, reject) => {
        const req = https.get(url, {
            headers: { 'User-Agent': UA, ...headers },
        }, (res) => {
            let body = '';
            res.on('data', (chunk) => body += chunk);
            res.on('end', () => {
                resolve({ statusCode: res.statusCode, headers: res.headers, body });
            });
        });
        req.on('error', reject);
        req.setTimeout(15000, () => req.destroy(new Error('Timeout')));
    });
}

// =========== Cookie/Crumb auth ===========
let cookies = '';
let crumb = '';
let crumbFailed = false;

async function refreshCrumb() {
    if (crumbFailed) return false;

    for (let attempt = 1; attempt <= 3; attempt++) {
        try {
            const initRes = await httpsGet('https://fc.yahoo.com/');
            const setCookie = initRes.headers['set-cookie'];
            if (setCookie) {
                const arr = Array.isArray(setCookie) ? setCookie : [setCookie];
                cookies = arr.map(c => c.split(';')[0]).join('; ');
            }

            const crumbRes = await httpsGet('https://query2.finance.yahoo.com/v1/test/getcrumb', { Cookie: cookies });
            if (crumbRes.statusCode === 429) {
                if (attempt === 3) {
                    console.log('  ⚠ Crumb rate limited, will use chart-only mode');
                    crumbFailed = true;
                    return false;
                }
                await sleep(attempt * 3000);
                continue;
            }
            if (crumbRes.statusCode === 200) {
                crumb = crumbRes.body;
                return true;
            }
            throw new Error(`Crumb status: ${crumbRes.statusCode}`);
        } catch (err) {
            if (attempt === 3) { crumbFailed = true; return false; }
            await sleep(attempt * 2000);
        }
    }
    return false;
}

async function yahooGet(url) {
    if (crumbFailed || !crumb) {
        const ok = await refreshCrumb();
        if (!ok) return null;
    }
    const sep = url.includes('?') ? '&' : '?';
    const fullUrl = `${url}${sep}crumb=${encodeURIComponent(crumb)}`;
    const res = await httpsGet(fullUrl, { Cookie: cookies });
    if (res.statusCode === 200) return JSON.parse(res.body);
    if (res.statusCode === 429) {
        crumbFailed = true;
        return null;
    }
    return null;
}

// =========== Load tickers ===========
function loadTickers() {
    const csvPath = path.join(__dirname, '..', 'python-service', 'vanguard.csv');
    const content = fs.readFileSync(csvPath, 'utf-8');
    const tickers = content.split(/\r?\n/)
        .map(l => l.trim())
        .filter(l => l.length > 0 && /^[A-Z]+$/.test(l) && !l.includes('/'));
    return [...new Set(tickers)].sort();
}

// =========== Chart endpoint (no auth, always works) ===========
async function getChartQuote(ticker) {
    try {
        const res = await httpsGet(`https://query1.finance.yahoo.com/v8/finance/chart/${ticker}?interval=1d&range=1d`);
        if (res.statusCode !== 200) return null;
        const data = JSON.parse(res.body);
        const meta = data.chart?.result?.[0]?.meta;
        if (!meta) return null;
        return {
            symbol: meta.symbol || ticker,
            price: meta.regularMarketPrice,
            fiftyTwoWeekHigh: meta.fiftyTwoWeekHigh,
            fiftyTwoWeekLow: meta.fiftyTwoWeekLow,
            volume: meta.regularMarketVolume,
            shortName: meta.shortName || ticker,
            longName: meta.longName || ticker,
        };
    } catch { return null; }
}

// =========== Main scan ===========
async function main() {
    const subsetArg = process.argv[2];
    const allTickers = loadTickers();
    const tickers = subsetArg ? allTickers.slice(0, parseInt(subsetArg)) : allTickers;

    console.log(`\n🔍 TickSignals Scanner`);
    console.log(`   ${tickers.length} tickers to scan (${allTickers.length} total in CSV)\n`);

    const startTime = Date.now();
    const stocks = [];

    // ===== Phase 1: Chart quotes (no auth, reliable) =====
    console.log('📊 Phase 1: Fetching price data via chart endpoint...');
    const chartMap = new Map();
    const CHART_CONCURRENCY = 10;

    for (let i = 0; i < tickers.length; i += CHART_CONCURRENCY) {
        const batch = tickers.slice(i, i + CHART_CONCURRENCY);

        const results = await Promise.allSettled(batch.map(t => getChartQuote(t)));

        for (const r of results) {
            if (r.status === 'fulfilled' && r.value) {
                chartMap.set(r.value.symbol, r.value);
            }
        }

        if ((i + CHART_CONCURRENCY) % 100 === 0 || i + CHART_CONCURRENCY >= tickers.length) {
            console.log(`  Progress: ${Math.min(i + CHART_CONCURRENCY, tickers.length)}/${tickers.length} — ${chartMap.size} valid`);
        }

        if (i + CHART_CONCURRENCY < tickers.length) await sleep(100);
    }

    console.log(`\n   ✓ Got ${chartMap.size} price quotes\n`);

    // ===== Phase 2: Try batch quotes for market cap + P/E (may fail if rate limited) =====
    const validTickers = Array.from(chartMap.keys());
    const quoteMap = new Map();

    const crumbOk = await refreshCrumb();
    if (crumbOk) {
        console.log('📈 Phase 2: Enriching with market cap, P/E via batch quotes...');
        const BATCH_SIZE = 50;

        for (let i = 0; i < validTickers.length; i += BATCH_SIZE) {
            if (crumbFailed) break;

            const batch = validTickers.slice(i, i + BATCH_SIZE);
            try {
                const data = await yahooGet(`https://query2.finance.yahoo.com/v7/finance/quote?symbols=${batch.join(',')}`);
                if (data) {
                    const results = data.quoteResponse?.result || [];
                    for (const q of results) quoteMap.set(q.symbol, q);
                }
            } catch { }

            if (i + BATCH_SIZE < validTickers.length) await sleep(200);
        }
        console.log(`   ✓ Got ${quoteMap.size} enriched quotes\n`);
    } else {
        console.log('⏭  Skipping batch quotes (rate limited), using chart data only\n');
    }

    // ===== Phase 3: Fundamentals via quoteSummary (may fail if rate limited) =====
    console.log('🔬 Phase 3: Fetching fundamentals...');
    const FUND_BATCH = 5;
    let fundCount = 0;
    let fundSkipped = 0;

    for (let i = 0; i < validTickers.length; i += FUND_BATCH) {
        const batch = validTickers.slice(i, i + FUND_BATCH);

        const fundResults = await Promise.allSettled(
            batch.map(async (ticker) => {
                if (crumbFailed) return null;
                return yahooGet(`https://query2.finance.yahoo.com/v10/finance/quoteSummary/${ticker}?modules=financialData,summaryDetail,assetProfile`);
            })
        );

        for (let j = 0; j < batch.length; j++) {
            const ticker = batch[j];
            const chart = chartMap.get(ticker);
            const quote = quoteMap.get(ticker);
            const fundResult = fundResults[j];
            const fundData = fundResult.status === 'fulfilled' ? fundResult.value : null;
            const fund = fundData?.quoteSummary?.result?.[0];

            const fin = fund?.financialData || {};
            const det = fund?.summaryDetail || {};
            const prof = fund?.assetProfile || {};

            if (fund) fundCount++;
            else fundSkipped++;

            const totalRevenue = fin.totalRevenue?.raw || null;
            const employees = prof.fullTimeEmployees || null;

            stocks.push({
                ticker,
                price: chart?.price || quote?.regularMarketPrice || null,
                market_cap: quote?.marketCap || null,
                pe_ratio: quote?.trailingPE || det?.trailingPE?.raw || null,
                roe_pct: fin.returnOnEquity?.raw != null ? fin.returnOnEquity.raw * 100 : null,
                debt_to_equity: fin.debtToEquity?.raw || null,
                gross_margin_pct: fin.grossMargins?.raw != null ? fin.grossMargins.raw * 100 : null,
                dividend_yield_pct: det.dividendYield?.raw != null ? det.dividendYield.raw * 100 : null,
                roa_pct: fin.returnOnAssets?.raw != null ? fin.returnOnAssets.raw * 100 : null,
                total_revenue: totalRevenue,
                revenue_per_employee: (totalRevenue && employees && employees > 0) ? totalRevenue / employees : null,
                sector: prof.sector || null,
                industry: prof.industry || null,
                company_name: chart?.shortName || chart?.longName || quote?.shortName || quote?.longName || ticker,
                fifty_two_week_high: chart?.fiftyTwoWeekHigh || quote?.fiftyTwoWeekHigh || null,
                fifty_two_week_low: chart?.fiftyTwoWeekLow || quote?.fiftyTwoWeekLow || null,
                beta: det.beta?.raw || null,
                updated_at: new Date().toISOString(),
            });
        }

        if ((i + FUND_BATCH) % 100 === 0 || i + FUND_BATCH >= validTickers.length) {
            console.log(`  Progress: ${Math.min(i + FUND_BATCH, validTickers.length)}/${validTickers.length} — ${fundCount} with fundamentals`);
        }

        if (i + FUND_BATCH < validTickers.length) await sleep(150);
    }

    // Save to JSON file
    const output = {
        stocks: stocks.sort((a, b) => (b.market_cap || 0) - (a.market_cap || 0)),
        totalStocks: stocks.length,
        lastUpdated: new Date().toISOString(),
        scanDuration: ((Date.now() - startTime) / 1000).toFixed(1),
    };

    fs.writeFileSync(DATA_FILE, JSON.stringify(output));

    const duration = ((Date.now() - startTime) / 1000).toFixed(1);
    console.log(`\n✅ Scan complete!`);
    console.log(`   ${stocks.length} stocks saved (${fundCount} with fundamentals, ${fundSkipped} chart-only)`);
    console.log(`   Duration: ${duration}s`);
    console.log(`   Output: ${DATA_FILE}\n`);
}

main().catch(err => {
    console.error('❌ Scan failed:', err.message);
    process.exit(1);
});

File: scripts/scan.py
#!/usr/bin/env python3
import sys
import os
import json
import time
import pandas as pd
import yfinance as yf
from datetime import datetime

# Path setup
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATA_DIR = os.path.join(BASE_DIR, 'data')
CSV_PATH = os.path.join(BASE_DIR, 'python-service', 'vanguard.csv')
CACHE_FILE = os.path.join(DATA_DIR, 'screener_cache.json')

if not os.path.exists(DATA_DIR):
    os.makedirs(DATA_DIR)

def load_tickers():
    try:
        with open(CSV_PATH, 'r', encoding='utf-8') as f:
            lines = f.read().splitlines()
        # Filter valid tickers - allow letters only, no slashes
        tickers = [l.strip() for l in lines if l.strip() and '/' not in l.strip() and l.strip().replace(' ', '').isalpha()]
        return sorted(list(set(tickers)))[:100]  # Limit to 100 for speed
    except Exception as e:
        print(f"Warning: Could not load tickers from {CSV_PATH}: {e}")
        return ['AAPL', 'MSFT', 'GOOGL', 'AMZN', 'TSLA']

def main():
    start_time = time.time()
    
    # Parse arguments
    subset = None
    if len(sys.argv) > 1:
        try:
            subset = int(sys.argv[1])
        except ValueError:
            pass

    all_tickers = load_tickers()
    if subset:
        tickers = all_tickers[:subset]
    else:
        # Limit to first 100 for testing speed if running full?
        # User said "stocks aren't getting pulled", so maybe run all or plenty.
        # But for robustness let's just run all requested.
        tickers = all_tickers

    # If list is empty
    if not tickers:
        print("No tickers found.")
        sys.exit(0)

    print(f"\n🔍 TickSignals Scanner (Python)")
    print(f"   {len(tickers)} tickers to scan ({len(all_tickers)} total in CSV)\n")

    stocks = []
    
    # Batch processing
    BATCH_SIZE = 50
    total_processed = 0

    for i in range(0, len(tickers), BATCH_SIZE):
        batch = tickers[i:i+BATCH_SIZE]
        print(f"   Processing batch {i} to {min(i+BATCH_SIZE, len(tickers))}...")

        try:
            # Download 1 year data to calculate 52w high/low
            # threads=True is default but let's be explicit
            data = yf.download(batch, period="1y", interval="1d", progress=False, group_by='ticker', auto_adjust=True, threads=True)
            
            current_date_iso = datetime.now().isoformat()

            for symbol in batch:
                try:
                    df = None
                    try:
                        # Try accessing as MultiIndex
                        df = data[symbol]
                    except KeyError:
                        # If failed, check if it's the only ticker and non-MultiIndex
                        if len(batch) == 1 and not isinstance(data.columns, pd.MultiIndex):
                            df = data
                        else:
                            # Symbol data missing
                            continue
                    
                    if df is None or df.empty:
                        continue
                        
                    # Get last valid row for current price
                    last_row = df.iloc[-1]
                    
                    # Check if 'Close' is NaN
                    if pd.isna(last_row['Close']):
                        continue
                        
                    current_price = float(last_row['Close'])
                    volume = int(last_row['Volume']) if not pd.isna(last_row['Volume']) else 0
                    
                    # Calculate 52-week stats
                    # 'High' and 'Low' columns
                    high_52 = float(df['High'].max())
                    low_52 = float(df['Low'].min())
                    
                    stock_data = {
                        "ticker": symbol,
                        "price": current_price,
                        "fifty_two_week_high": high_52,
                        "fifty_two_week_low": low_52,
                        "market_cap": None, 
                        "pe_ratio": None,
                        "company_name": symbol, # Fallback to symbol
                        "updated_at": current_date_iso
                    }
                    stocks.append(stock_data)

                except Exception as e:
                    # print(f"Error processing {symbol}: {e}")
                    pass
            
            total_processed += len(batch)
            print(f"   Progress: {len(stocks)} valid stocks so far")
            
        except Exception as e:
            print(f"   Batch error: {e}")

    # Sort by price descending (like original script)
    stocks.sort(key=lambda x: x.get('price', 0), reverse=True)

    duration = round(time.time() - start_time, 1)

    output = {
        "stocks": stocks,
        "totalStocks": len(stocks),
        "lastUpdated": datetime.now().isoformat(),
        "scanDuration": str(duration)
    }

    try:
        with open(CACHE_FILE, 'w') as f:
            json.dump(output, f, indent=2)
        
        print(f"\n✅ Scan complete!")
        print(f"   {len(stocks)} stocks saved to {CACHE_FILE}")
        print(f"   Duration: {duration}s\n")
        
    except Exception as e:
        print(f"❌ Error writing cache file: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
